OpenRA.Mods.RA/Activities/TeslaBoost.cs

```diff
@@ -0,0 +1,138 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.Activities;
+using OpenRA.GameRules;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Mods.RA.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.RA.Activities
+{
+	public class TeslaBoost : Activity
+	{
+		readonly Actor target;
+
+		readonly WRange range;
+
+		readonly UpgradeManager um;
+		readonly string[] upgrades;
+		readonly int grantMultiple;
+
+		readonly bool pursueTarget;
+
+		readonly IFacing facing;
+
+		readonly int intervalTicks;
+		readonly Armament arm;
+		readonly Target ttarget;
+
+		bool upgraded = false;
+		int ticks = 0;
+
+		public TeslaBoost(Actor self, Actor target, bool pursueTarget)
+		{
+			this.target = target;
+
+			var uti = self.Info.Traits.Get<GivesTeslaBoostInfo>();
+
+			range = uti.Range;
+
+			um = target.TraitOrDefault<UpgradeManager>();
+			upgrades = uti.Upgrades;
+			grantMultiple = uti.GrantMultiple;
+
+			this.pursueTarget = pursueTarget;
+
+			facing = self.TraitOrDefault<IFacing>();
+
+			intervalTicks = uti.IntervalTicks;
+			arm = self.Trait<AttackBase>().Armaments.First(a => a.Info.Name == uti.Armament);
+			ttarget = Target.FromActor(target);
+		}
+
+		public override Activity Tick(Actor self)
+		{
+			// Should the target for the boost be destroyed, cancel the activity
+			if (!target.IsInWorld)
+				return NextActivity;
+
+			// Should the activity be canceled or target moves out of range, cancel the activity
+			if (IsCanceled || (self.CenterPosition - target.CenterPosition).LengthSquared > range.RangeSquared)
+			{
+				if (upgraded)
+				{
+					upgraded = false;
+					if (um != null)
+						foreach (var upgrade in upgrades)
+							for (var u = 0; u < grantMultiple; u++)
```

Comment:
If `grantMultiple` is to be used as a cap, the it should have a name like `maxGrantCount` instead of one that implies boolean value.